### PR TITLE
Parameterizations of sub-grid topographic effects on solar radiation in ELM

### PR DESCRIPTION
**Pull request by daleihao**
A well-validated sub-grid topographic (TOP) parameterization is added in ELM
that quantifies the effects of sub-grid topography on solar radiation flux
including the shadow effects and multi-scattering between adjacent terrain.
The new parameterization can be turned on by `use_top_solar_rad = .true.`
A test for the new parameterization is added.

[BFB]

### CODE REVIEW

**Comment by bishtgautam** [comment]
@rljacob After additional analysis, I believe the test can be changed to a regular ERS test. Does that sound okay?

**Comment by rljacob** [comment]
You mean a 7 day ERS test?  That's fine unless theres a reason it need to run for a month.

**Review by rljacob — APPROVED** [review]

**Comment by rljacob** [comment]
The ERS test for this feature is failing on gnu and on pgi:
ERS.f09_f09.IELM.mappy_gnu.elm-solar_rad
ERS.f09_f09.IELM.compy_pgi.elm-solar_rad

SUMMARY of cprnc (from compy)
 A total number of    193 fields were compared
          of which     37 had non-zero differences
               and      0 had differences in fill patterns
 A total number of      0 fields could not be analyzed
 A total number of      0 fields on file 1 were not found on file2.
  diff_test: the two files seem to be DIFFERENT

oddly, its passing on Chrysalis with intel.

@bishtgautam I force-pushed next to remove your fix from this morning since we may have to revert the whole PR.  This way, you'll only have to revert one thing from next.

**Comment by bishtgautam** [comment]
Thanks @rljacob for the force-pushed

File: components/elm/src/biogeophys/SurfaceAlbedoType.F90 [hunk 1]
```diff
@@ -265,6 +270,11 @@ subroutine InitAllocate(this, bounds)
     allocate(this%albd_patch         (begp:endp,numrad))       ; this%albd_patch         (:,:) =spval
     allocate(this%albi_patch         (begp:endp,numrad))       ; this%albi_patch         (:,:) =spval
 
+    !!=== TOP solar radiation parameterization =====
+	allocate(this%fd_top_adjust      (begp:endp,numrad))       ; this%fd_top_adjust      (:,:) = spval
+	allocate(this%fi_top_adjust      (begp:endp,numrad))       ; this%fi_top_adjust      (:,:) = spval
+    !!=== End
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoType.F90:276]: delete the comments

File: components/elm/src/biogeophys/SurfaceAlbedoType.F90 [hunk 2]
```diff
@@ -339,7 +349,19 @@ subroutine InitHistory(this, bounds)
     call hist_addfld2d (fname='ALBI', units='proportion', type2d='numrad', &
          avgflag='A', long_name='surface albedo (indirect)', &
          ptr_patch=this%albi_patch, default='inactive', c2l_scale_type='urbanf')
-
+	
+    !!=== TOP solar radiation parameterization =====
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoType.F90:353]: delete the comments

File: components/elm/src/biogeophys/SurfaceAlbedoType.F90 [hunk 3]
```diff
@@ -339,7 +349,19 @@ subroutine InitHistory(this, bounds)
...
+    this%fd_top_adjust(begp:endp,:) = spval
+    call hist_addfld2d (fname='fd_top_adjust', units='none', type2d='numrad', &
+         avgflag='A', long_name='fd_top_adjust', &
+         ptr_patch=this%fd_top_adjust)
+		 
+	this%fi_top_adjust(begp:endp,:) = spval
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoType.F90:359]: indent it

File: components/elm/src/biogeophys/SurfaceAlbedoType.F90 [hunk 4]
```diff
@@ -387,6 +409,11 @@ subroutine InitCold(this, bounds)
     this%ftid_patch     (begp:endp, :) = 0.0_r8
     this%ftii_patch     (begp:endp, :) = 1.0_r8
 
+    !!=== TOP solar radiation parameterization =====
+    this%fd_top_adjust  (begp:endp, :) = 1.0_r8
+	this%fi_top_adjust  (begp:endp, :) = 1.0_r8
+    !!=== End
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoType.F90:415]: delete the comments and indent it

File: components/elm/src/data_types/GridcellType.F90 [hunk 5]
```diff
@@ -46,7 +46,15 @@ module GridcellType
      integer , pointer :: pftf         (:) => null() ! ending pft index for each gridcell
      integer , pointer :: npfts        (:) => null() ! number of patches for each gridcell
 
-     ! Physical properties
+     !!=== TOP solar radiation parameterization =====
```
Comment (bishtgautam) [inline components/elm/src/data_types/GridcellType.F90:49]: delete it

File: components/elm/src/data_types/GridcellType.F90 [hunk 6]
```diff
@@ -46,7 +46,15 @@ module GridcellType
...
+     real(r8), pointer :: stdev_elev   (:) => null()        ! standard deviation of elevation within a gridcell
+     real(r8), pointer :: sky_view     (:) => null()         ! mean of (sky view factor / cos(slope))
+     real(r8), pointer :: terrain_config (:) => null()    ! mean of (terrain configuration factor / cos(slope))
+     real(r8), pointer :: sinsl_cosas  (:) => null()     ! sin(slope)*cos(aspect) / cos(slope)
+     real(r8), pointer :: sinsl_sinas  (:) => null()     ! sin(slope)*sin(aspect) / cos(slope)
+     !!=== End
```
Comment (bishtgautam) [inline components/elm/src/data_types/GridcellType.F90:55]: delete it

File: components/elm/src/data_types/GridcellType.F90 [hunk 7]
```diff
@@ -103,7 +111,14 @@ subroutine grc_pp_init(this, begg, endg)
     allocate(this%pfti      (begg:endg)) ; this%pfti      (:) = ispval
     allocate(this%pftf      (begg:endg)) ; this%pftf      (:) = ispval
     allocate(this%npfts     (begg:endg)) ; this%npfts     (:) = ispval
-
+    !!=== TOP solar radiation parameterization =====
```
Comment (bishtgautam) [inline components/elm/src/data_types/GridcellType.F90:114]: Delete it

File: components/elm/src/data_types/GridcellType.F90 [hunk 8]
```diff
@@ -103,7 +111,14 @@ subroutine grc_pp_init(this, begg, endg)
...
+    allocate(this%stdev_elev(begg:endg)) ; this%stdev_elev(:) = ispval        ! standard deviation of elevation within a gridcell
+    allocate(this%sky_view  (begg:endg)) ; this%sky_view  (:) = ispval        ! mean of (sky view factor / cos(slope))
+    allocate(this%terrain_config(begg:endg)) ; this%terrain_config(:) = ispval! mean of (terrain configuration factor / cos(slope))
+    allocate(this%sinsl_cosas(begg:endg)) ; this%sinsl_cosas(:) = ispval      ! sin(slope)*cos(aspect) / cos(slope)
+    allocate(this%sinsl_sinas(begg:endg)) ; this%sinsl_sinas(:) = ispval      ! sin(slope)*sin(aspect) / cos(slope)
+    !!=== End
```
Comment (bishtgautam) [inline components/elm/src/data_types/GridcellType.F90:120]: Delete it

File: components/elm/src/main/controlMod.F90 [hunk 9]
```diff
@@ -303,6 +304,11 @@ subroutine control_init( )
     namelist /elm_inparm/ &
          use_erosion, ero_ccycle
 
+    ! === TOP solar radiation parameterization options  !
+    namelist /elm_inparm/ &
+         use_top_solar_rad
+    ! End
```
Comment (bishtgautam) [inline components/elm/src/main/controlMod.F90:310]: Delete the comments

File: components/elm/src/main/controlMod.F90 [hunk 10]
```diff
@@ -964,6 +971,13 @@ subroutine control_print ()
     else
        write(iulog,*) '   atm topographic data = ',trim(fatmtopo)
     end if
+    
+    if (use_top_solar_rad) then  ! TOP solar radiation parameterization
+        write(iulog,*) '  use TOP solar radiation parameterization instead of PP'
+    else
+        write(iulog,*) '   use_top_solar_rad is False, so do not run TOP solar radiation parameterization'
+    end if  ! End
```
Comment (bishtgautam) [inline components/elm/src/main/controlMod.F90:979]: Delete `! TOP solar radiation parameterization` and `! End`

File: components/elm/src/main/elm_initializeMod.F90 [hunk 11]
```diff
@@ -221,6 +222,16 @@ subroutine initialize1( )
           call shr_sys_flush(iulog)
        endif
        call surfrd_get_topo(ldomain, flndtopo)  
+    endif    
+    
+    
+    ! === Get topography parameters for TOP solar radiation parameterization
```
Comment (bishtgautam) [inline components/elm/src/main/elm_initializeMod.F90:228]: delet it

File: components/elm/src/main/surfrdMod.F90 [hunk 12]
```diff
@@ -1463,4 +1464,112 @@ subroutine surfrd_topounit_data(begg, endg, lfsurdat)
     
   end subroutine surfrd_topounit_data
 
+
+!-----------------------------------------------------------------------
+!!! TOP solar radiation parameterization
+! !IROUTINE: surfrd_get_topo_for_solar_rad
```
Comment (bishtgautam) [inline components/elm/src/main/surfrdMod.F90:1470]: delete it

File: components/elm/src/utils/domainMod.F90 [hunk 13]
```diff
@@ -45,6 +45,14 @@ module domainMod
      logical          :: set        ! flag to check if domain is set
      logical          :: decomped   ! decomposed locally or global copy
 
+     ! === TOP solar radiation parameterization
```
Comment (bishtgautam) [inline components/elm/src/utils/domainMod.F90:48]: delete it

File: components/elm/src/utils/domainMod.F90 [hunk 14]
```diff
@@ -45,6 +45,14 @@ module domainMod
...
+     real(r8),pointer :: stdev_elev(:)     ! standard deviation of elevation within a gridcell
+     real(r8),pointer :: sky_view(:)       ! mean of (sky view factor / cos(slope))  ! 3D-RT
+     real(r8),pointer :: terrain_config(:) ! mean of (terrain configuration factor / cos(slope))
+     real(r8),pointer :: sinsl_cosas(:)    ! sin(slope)*cos(aspect) / cos(slope)
+     real(r8),pointer :: sinsl_sinas(:)    ! sin(slope)*sin(aspect) / cos(slope)
+     ! === End
```
Comment (bishtgautam) [inline components/elm/src/utils/domainMod.F90:54]: delete it

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 [hunk 15]
```diff
@@ -939,7 +944,8 @@ subroutine SurfaceAlbedo(bounds,     &
                coszen_patch(bounds%begp:bounds%endp), &
                rho(bounds%begp:bounds%endp, :), &
                tau(bounds%begp:bounds%endp, :), &
-               canopystate_vars, surfalb_vars)
+               canopystate_vars, surfalb_vars, &
+               nextsw_cday, declinp1) ! For TOP solar radiation parameterization
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:948]: delete `! For TOP solar radiation parameterization`

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 [hunk 16]
```diff
@@ -1566,7 +1610,7 @@ subroutine TwoStream(bounds, &
 
                 d_ftii = -h * h9 * s1 + h * h10 / s1 + dh9 * s1 + dh10 / s1
                 d_fabi = -(dh7+dh8) - (1._r8-albgri(c,ib))*d_ftii
-                d_fabi_sun = (1._r8 - omega(p,ib)) / avmu(p) * (da1 + da2)
+                d_fabi_sun = (1._r8 - omega(p,ib)) / avmu(p) * (da1 + da2) * fi_top_adjust(p,ib) ! revised by Dalei Hao for TOP
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:1613]: delete `revised by Dalei Hao for TOP`

**Review by bishtgautam — CHANGES_REQUESTED** [review]
@daleihao The PR needs code cleanup related to comments and indentation.

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 [hunk 17]
```diff
@@ -1589,5 +1619,559 @@ subroutine TwoStream(bounds, &
...
+    data coeff_dif(:,6) /7.015E-7, 3.298E+0, 2.543E-1, -3.547E+0/
+    data coeff_dif(:,7) /6.359E-7, 2.984E+0,        0, -2.984E+0/
+
+    data coeff_rdif(:,1) / 1.493E-1, 1.621E-1, -1.483E-1/
+    data coeff_rdif(:,2) / 1.462E-1, 1.654E-1, -1.454E-1/
+    data coeff_rdif(:,3) / 1.454E-1, 1.673E-1, -1.446E-1/
+    data coeff_rdif(:,4) / 1.465E-1, 1.683E-1, -1.457E-1/
+    data coeff_rdif(:,5) / 1.443E-1, 1.682E-1, -1.435E-1/
+    data coeff_rdif(:,6) / 1.446E-1, 1.686E-1, -1.439E-1/
+    data coeff_rdif(:,7) /-3.427E-6, 1.576E-1,  1.199E-3/
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:1708]: need to add `_r8` to all numbers i.e. `2.045E+1_r8`

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 [hunk 18]
```diff
@@ -1589,5 +1619,559 @@ subroutine TwoStream(bounds, &
...
+    data coeff_dir(:,6) /2.977E-3, 9.959E-1, -9.990E-1/
+    data coeff_dir(:,7) /8.347E-3,        0, -8.393E-3/
+
+    data coeff_rdir(:,1) / 2.351E-1, 1.590E-1, -2.332E-1/
+    data coeff_rdir(:,2) / 1.368E-1, 1.642E-1, -1.358E-1/
+    data coeff_rdir(:,3) / 1.254E-1, 1.653E-1, -1.247E-1/
+    data coeff_rdir(:,4) / 1.274E-1, 1.635E-1, -1.267E-1/
+    data coeff_rdir(:,5) / 1.314E-1, 1.623E-1, -1.307E-1/
+    data coeff_rdir(:,6) / 1.359E-1, 1.620E-1, -1.352E-1/
+    data coeff_rdir(:,7) /-4.463E-6, 1.556E-1,  1.287E-3/
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:1917]: similarly, `_r8` is needed

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 [hunk 19]
```diff
@@ -1589,5 +1619,559 @@ subroutine TwoStream(bounds, &
...
+    data coeff_dif(:,6) /7.015E-7, 3.298E+0, 2.543E-1, -3.547E+0/
+    data coeff_dif(:,7) /6.359E-7, 2.984E+0,        0, -2.984E+0/
+
+    data coeff_rdif(:,1) / 1.493E-1, 1.621E-1, -1.483E-1/
+    data coeff_rdif(:,2) / 1.462E-1, 1.654E-1, -1.454E-1/
+    data coeff_rdif(:,3) / 1.454E-1, 1.673E-1, -1.446E-1/
+    data coeff_rdif(:,4) / 1.465E-1, 1.683E-1, -1.457E-1/
+    data coeff_rdif(:,5) / 1.443E-1, 1.682E-1, -1.435E-1/
+    data coeff_rdif(:,6) / 1.446E-1, 1.686E-1, -1.439E-1/
+    data coeff_rdif(:,7) /-3.427E-6, 1.576E-1,  1.199E-3/
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/SurfaceAlbedoMod.F90:2078]: `_r8` is needed

**Review by bishtgautam — COMMENTED** [review]
@daleihao The changes look great. I have have few additional minor changes that I forgot to ask you to fix in my first review.

File: components/elm/src/biogeophys/SurfaceAlbedoMod.F90 (hunk 17 above)
Comment (daleihao) [reply components/elm/src/biogeophys/SurfaceAlbedoMod.F90:1708]: I see. Will revise them

**Review by bishtgautam — APPROVED** [review]

**Comment by rljacob** [comment]
This was reverted from next because it had unexpected diffs.  Please fix.

**Comment by rljacob** [comment]
@daleihao have you fixed the unexpected diffs?

**Comment by daleihao** [comment]
> @daleihao have you fixed the unexpected diffs?

I have checked the code. All the tests passed the regression test. Hope it works now.

**Comment by rljacob** [comment]
@bishtgautam ready to try merging this again?

**Comment by bishtgautam** [comment]
@rljacob If next is open today, I will merge this PR today.

**Comment by rljacob** [comment]
@bishtgautam next is open today.

**Comment by bishtgautam** [comment]
Merged to next

File: components/elm/cime_config/testdefs/testmods_dirs/elm/solar_rad/shell_commands [hunk 20]
```diff
@@ -0,0 +1 @@
+./xmlchange NTASKS=48
```
Comment (amametjanov) [inline components/elm/cime_config/testdefs/testmods_dirs/elm/solar_rad/shell_commands:1]: @daleihao is there a reason to set NTASKS=48 ? With this, `ERS.f09_f09.IELM.ascent_pgi.elm-solar_rad` is erroring out in 2nd/restart run with memory corruption
```
2: 14: 0: ALLOCATE: 2356269769344 bytes requested; not enough memory
2: 42: 0: ALLOCATE: 5442822862464 bytes requested; not enough memory
```
Changing NTASKS to default 1-node with 84 NTASKS lets both initial and restart runs complete BFB. Okay to remove this mod?

**Comment by daleihao** [comment]
It doesn't matter what  NTASKs is. It is OK to remove it.

Best,
-Dalei

Azamat Mametjanov ***@***.***> 于2022年4月8日周五 11:27写道：

File: components/elm/cime_config/testdefs/testmods_dirs/elm/solar_rad/shell_commands (hunk 20 above)
Comment (daleihao) [reply components/elm/cime_config/testdefs/testmods_dirs/elm/solar_rad/shell_commands:1]: It doesn't matter what  NTASKs is. It is OK to remove it.

**Pull request merged**